src/MonitorProfiler/ProbeInstrumentation/ProbeInstrumentation.cpp

```diff
@@ -0,0 +1,300 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "corhlpr.h"
+#include "ProbeInstrumentation.h"
+
+using namespace std;
+
+#define IfFailLogRet(EXPR) IfFailLogRet_(m_pLogger, EXPR)
+
+ProbeInstrumentation::ProbeInstrumentation(const shared_ptr<ILogger>& logger, ICorProfilerInfo12* profilerInfo) :
+    m_pCorProfilerInfo(profilerInfo),
+    m_pLogger(logger),
+    m_probeFunctionId(0),
+    m_pAssemblyProbePrep(nullptr)
+{
+}
+
+HRESULT ProbeInstrumentation::RegisterFunctionProbe(FunctionID enterProbeId)
+{
+    lock_guard<mutex> lock(m_probePinningMutex);
+
+    if (HasPinnedProbe())
+    {
+        // Probes have already been pinned.
```

Comment:
Not sure whether it makes sense in the current context, but given that we log down below when we _do_ receive probes, would we want to log when probes were already pinned? Or would you just note the absence of that message?